### PR TITLE
Switch from Paths to Sources

### PR DESCRIPTION
**Pull request by ArquintL**
This PR adapts Gobra to take a vector of `Source` instead of `Path` as input. This provides greater flexibility as a `Source` can either represent a file or simply some string. In particular, it allows Gobra Server to invoke Gobra with some (file)content that is not necessarily stored in local files.
Not however that this does not immediately enables Gobra Server to run on a remote machine as the entire import resolution mechanism still relies on the local filesystem. A proper abstraction would need to be implemented (e.g. with callbacks to query the directory structure and/or to retrieve files).

### CODE REVIEW

File: src/main/scala/viper/gobra/frontend/Parser.scala [hunk 1]
```diff
@@ -129,37 +112,53 @@ object Parser {
...
-      }
+    def samePackage(programs: Vector[PProgram]): Either[Vector[ParserError], Vector[PProgram]] = {
+      require(programs.nonEmpty)
+      val pkgName = programs.head.packageClause.id.name
+      val differingPkgNameMsgs = programs.flatMap(p =>
+        error(
+          p.packageClause,
+          s"Files have differing package clauses, expected $pkgName but got ${p.packageClause.id.name}",
+          p.packageClause.id.name != pkgName))
+      println(s"samePackage msgs: ${differingPkgNameMsgs}")
```
Comment (jcp19) [inline src/main/scala/viper/gobra/frontend/Parser.scala]: is this print supposed to be here or is this just a debugging device?

File: src/main/scala/viper/gobra/frontend/Parser.scala [hunk 2]
```diff
@@ -129,37 +112,53 @@ object Parser {
...
+    val res = for {
+      // check that each of the parsed programs has the same package clause. If not, the algorithm collecting all files
+      // of the same package has failed or users have entered an invalid collection of inputs
+      programs <- isErrorFree(parserResults)
+      programs <- samePackage(programs)
+      pkg <- makePackage(programs)
+    } yield pkg
+    // report potential errors:
+    res.left.map(errors => {
+      println(s"errors: ${errors}")
```
Comment (jcp19) [inline src/main/scala/viper/gobra/frontend/Parser.scala]: same as before

**Review by jcp19 — APPROVED** [review]
LGTM, I just have a few minor comments

File: src/main/scala/viper/gobra/ast/frontend/Ast.scala [hunk 3]
```diff
@@ -76,7 +75,7 @@ class PositionManager(val positions: Positions) extends Messaging(positions) {
     val path = start.source match {
       case FileSource(filename, _) => Paths.get(filename)
       case FromFileSource(path, _) => path
-      case _ => ???
+      case StringSource(_, _) => null // StringSource stores some name but there is no guarantee that it corresponds to a valid file path
```
[the remark this answers is not in the document]
Comment (ArquintL) [reply src/main/scala/viper/gobra/ast/frontend/Ast.scala]: I've left it at `???` and mentioned multiple ways how to potentially fix this in the future as I currently do not need it for supporting packages in the IDE

File: src/main/scala/viper/gobra/frontend/Source.scala [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+      case FromFileSource(path, _) => path
+      case StringSource(_, _) =>
+        // StringSource stores some name but there is no guarantee that it corresponds to a valid file path
+        // there are the following ideas:
+        // - create a file adhoc if a filepath is really necessary (in the spirit of `StringSource.useAsFile`)
+        // - adapt Silver to not require a Path object but some abstraction
+        // - create an implementation implementing Path that simply stores the StringSource and leaves all functions
+        //    unimplemented. This works as long as nobody relies on any operations available for paths. It's also bad
+        //    style but at least any use fails in an understandable way (as opposed to simply returning `null` here)
+      ???
```
Comment (jcp19) [inline src/main/scala/viper/gobra/frontend/Source.scala:40]: This is a lot safer, thanks!

File: src/main/scala/viper/gobra/reporting/Message.scala [hunk 5]
```diff
@@ -46,7 +46,12 @@ case class GobraOverallFailureMessage(verifier: String, result: VerifierResult)
     s"failure=${result.toString})"
 }
 
-case class GobraEntitySuccessMessage(verifier: String, concerning: Source.Verifier.Info) extends GobraVerificationResultMessage {
+sealed trait GobraEntityResultMessage extends GobraVerificationResultMessage {
```
Comment (jcp19) [inline src/main/scala/viper/gobra/reporting/Message.scala:49]: What is the purpose of storing `entity` here? I might have missed something but it does not seem to be used by any of its subcalsses

**Review by jcp19 — APPROVED** [review]

File: src/main/scala/viper/gobra/reporting/Message.scala (hunk 5 above)
Comment (ArquintL) [reply src/main/scala/viper/gobra/reporting/Message.scala:49]: It allows Gobra Server to figure out to which entity a message belongs. In particular, it keeps track of the progress by having a list of entities extracted from the Viper AST and checking whether an entity message belongs to any of these

Comment (jcp19) [reply src/main/scala/viper/gobra/reporting/Message.scala:49]: I see, you use it to keep track of the progress in the GobraIDE

**Pull request merged**